pip/pip-338.md

```diff
@@ -0,0 +1,172 @@
+
+But the challenge arises when it considers the same internalListener for lookup requests redirects. This may result in returning an unresolvable broker service address to the clients.
+It may not be possible to pass the listenerName from the clients, which consequently may lead to cluster downtime.
+
+it might not be feasible to have the listerner config at the client side in every tech stack or connector.
+Also,  Currently, there is no option to pass the listener while using the admin APIs. As admin APIs can be called from an external network as well, the use of the internal listener’s broker service URL can lead all admin operations to get affected.
+
+Moreover, as per current code provided above, even if the client provides a listenerName/listener header, the redirect urls are taken from the internal listener or the first listener with http protocol which may not be a client-resolvable http address.
+
+# Proposed Solution
+To fix inconsistency with all protocol’s service url, we need to include all broker addresses associated with the listener name while returning the lookupResult, rather than solely the service URL
+
+Also, Introduce a broker-level property to determine the appropriate listener for lookup calls if listenerName is missing from the client request. This can help deal with the side effects of not having listerName on the client side when multiple advertisedListeners are being used.
+
+Also, this can help to determine the broker service URL that needs to be returned in case of lookup redirects and also for using Pulsar admin API calls from outside the network.
+
+This will help in having a more transparent listener identifier for internal and external communications.
+
+
+# Design & Implementation Details
+## Approach
+**1. Fix Code Flow - Inconsistency with listener usage with protocols other than pulsar and pulsar+ssl** 
+
+Currently, lookup result return the broker service url corresponding to the internal listerner in case of http or https protocol urls. To address the issue in the workflow, it is necessary to include all broker addresses associated with the listener name, rather than solely the service URL. There can be a check if all broker addresses corresponding to the given listener is null, then use the default listener(discussed below)
+
+**2. Only return the service URL corresponding to the lookupListener**
+
+This approach introduces one new configuration `lookupListenerName` in the broker.conf. The `lookupListenerName` is used to specify which service URL should be returned to the clients from the lookup requests or redirects. `lookupListenerName` should be present in the `advertisedListeners` list.
```

Comment:
This doesn't make sense to me. The `internalListenerName` is already a default listener. I don't see how this would solve the gap in the solution which is that all Pulsar Admin REST APIs don't have the redirects covered with the selected listener name.

Reply:
the users will configure an external, client resolvable listener name as the `lookupListnerName` (ex - an IP). It will be different from `internalListerName` in case the internal address (ex - pod fqn) is non-resolvable outside the cluster/k8s layer.

Reply:
How would you know what to use, the "internalListenerName" or the "lookupListenerName"? I think that "internalListenerName" is badly named, it's essentially the default listener name. From that perspective, the "lookupListenerName" is conflicting since it's another default.

The problem is that lookups are needed also within the cluster and perhaps for multiple different listeners. The design of PIP-95 is that there isn't a single "external" listener. I think it's better to keep the current PIP-95 design and perhaps extend it to include `bindAddresses` support also for the web endpoints. 

That way it would be possible to have a different default listener name for different bind address/port as it is already possible for the Pulsar binary protocol. That was a scope limitation of PIP-95, mentioned also [in this comment](https://github.com/apache/pulsar/issues/12040#issuecomment-947114322).

Reply:
> How would you know what to use, the "internalListenerName" or the "lookupListenerName"? I think that "internalListenerName" is badly named, it's essentially the default listener name. From that perspective, the "lookupListenerName" is conflicting since it's another default.
> 

Any documentation supporting this idea of internallistenerName being default? Because the way it is used currently, it makes it evident that the address represented by this listener is supposed to be an intenal usage thing. 

The core issue arises when your deterministic sticky named address is not resolvable outside the cluster. If we ignore that problem, then this is a non issue. Or in other words, the problem arises when we want to use deterministic sticky addresses to brokers as identifier rather than its IP address. One of the main reason to do this is the fact that this address will be used to identify a broker wrt bundle ownership i.e. even a single pod restart would lead to removal of that broker and creation of a "new" broker in the cluster. Another use case of having a static deterministic broker address is for the regex used in namespace isolation policies.

Generally, these static addresses (stateful pod name) are not resolvable outside the k8s cluster.

Thus, we have to set out advertised address as this static, non-client-resolvable address as we are using isolation groups feature and for the other reason mentioned above as well.

Now as soon as we do that, we have to set the same address in the list of advertised addresses as well, that too, as the internal listener address other wise the check at this line breaks in `searchForCandidateBroker` which leads to lookup timeouts:

https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java#L533

https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java#L550

because `isBrokerActive` does a check of <internal address> in a list of <advertised addresses>. So we have to set `internalListenerName` and ensure that the internal listener is in the same format (deterministic non-IP based host name) as the advertised listener.

So, now, as per implementation itself, at this point, internalListenerName cannot be assumed as the default listener as its something only resolveable inside the pulsar cluster/k8s namespace.


> The problem is that lookups are needed also within the cluster and perhaps for multiple different listeners. The design of PIP-95 is that there isn't a single "external" listener. I think it's better to keep the current PIP-95 design and perhaps extend it to include `bindAddresses` support also for the web endpoints.
> 
> That way it would be possible to have a different default listener name for different bind address/port as it is already possible for the Pulsar binary protocol. That was a scope limitation of PIP-95, mentioned also [in this comment](https://github.com/apache/pulsar/issues/12040#issuecomment-947114322).


Could you run us through how that would work for the case when a normal producer client does an http based lookup call only to use the native protocol response from it and to also ensure that redirect in case of 307 response of that lookup call also works correctly? 
We have to consider the fact that due to many security/audit reasons, there might only be a single http based load balancer at port 80 possible for the brokers which may only talk to a single broker-port of 8080.

Reply:
> The core issue arises when your deterministic sticky named address is not resolvable outside the cluster. If we ignore that problem, then this is a non issue. Or in other words, the problem arises when we want to use deterministic sticky addresses to brokers as identifier rather than its IP address. 

It would be expected that there would be a way to address that. There are simply bugs and gaps in the current solution as it has been discussed and I think that we are pretty close in getting to a consensus. 

> Any documentation supporting this idea of internallistenerName being default? Because the way it is used currently, it makes it evident that the address represented by this listener is supposed to be an intenal usage thing.

It's referred to as "internal listener", but in the implementation, it is used as the default. For example 
there is "primary channel" in the code comment: https://github.com/apache/pulsar/blob/5df97b41a431a714a448c198c312e98aa4a084e8/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L543-L552

Since the "internalListenerName" is set to the first listener if it is unset, there's always going to be one listener that is the "internal listener". In many ways it could be seen as a default.

> Now as soon as we do that, we have to set the same address in the list of advertised addresses as well, that too, as the internal listener address other wise the check at this line breaks in searchForCandidateBroker which leads to lookup timeouts:

The PRs #21894 & #21937 fixed the issue #21897 and there's no coupling from the brokerId to the "advertised listeners" any more. The fixes have been backported all the way to branch-3.0 .

Reply:
yes, after your PRs the issue of `finding advertised listener in list of internal listeners` is fixed. But this regression is introduced in 2.10 . We are upgrading from 2.9 to 2.10.
The only reason it would work post 3.0 is because we will stop setting an `internalListenerName` all together. Because as soon as we start setting that, we will once again come into the issue of all the inconsistencies mentioned in this PIP and the _need_ to take the listener name from incoming lookup/admin calls.

Reply:
> But this regression is introduced in 2.10 . We are upgrading from 2.9 to 2.10.

According to the current release policy, there's no active maintenance for 2.10 branch in the Apache Pulsar project, the dev mailing list thread is https://lists.apache.org/thread/n5zfhohvk7olodfx2gksxk9n5f2pt3h8 . The Apache Pulsar community and Apache Pulsar PMC could change this, but this is the current state.
It would be better to target 3.0.x which is the current LTS release and will be actively supported until May 2025.
I would assume that it's possible to upgrade from 2.9 to 3.0.x , but that obviously requires testing in any case.

Reply:
I checked this in slack, and we will go from 2.10 to 3.0.x directly, but we were already halfway done with 2.9 to 2.10 upgrade and only in production found this bug.

Currently there is no documentation on upgrade paths at all. All previous conversations with SN or on slack were indicating to do upgrades one version at a time.

Reply:
> The only reason it would work post 3.0 is because we will stop setting an `internalListenerName` all together. Because as soon as we start setting that, we will once again come into the issue of all the inconsistencies mentioned in this PIP and the _need_ to take the listener name from incoming lookup/admin calls.

This still holds true.

Reply:
> Currently there is no documentation on upgrade paths at all. All previous conversations with SN or on slack were indicating to do upgrades one version at a time.

That's true, however that's a general rule in thumb. In this case since 3.0.x is the oldest actively supported, it makes sense to target it. If an upgrade path works between 2 versions in tests & experiments, there shouldn't be a reason why it couldn't be used. Obviously it might require some more care in validating if rollbacks are possible.